### PR TITLE
Adds dataframe style logical plan builders

### PR DESCRIPTION
**Pull request by RCHowell**
## Relevant Issues

https://github.com/partiql/partiql-lang-kotlin/pull/1535

## Description

This is a follow-up to #1535 which initialized the V1 plan builders. This uses a fluent/dataframe style API to construct logical plans. It consolidates where the standard implementations live. The plan builders are deferred builders with a factory for providing node implementations.

**Example**
```kotlin

val table = MyTable(..)

// SELECT * FROM MyTable WHERE x < 3
val plan = RelBuilder
              .scan(table)
              .filter(RexBuilder.col("x").lt(RexBuilder.lit(3)))
              .build()
```

**Ideally we can have a large set of top-level static functions to make construction less verbose – combined with the 'with' or 'use' syntax**

## Other Information
- Updated Unreleased Section in CHANGELOG: **NO**
- Any backward-incompatible changes? **NO**
- Any new external dependencies? **NO**
- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **YES**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rex/Rex.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import org.partiql.types.PType
+
+/**
+ * TODO DOCUMENTATION
+ */
+public interface Rex {
+
+    public fun getType(): PType
+
+    public fun getOperands(): List<Rex>
```
Comment (johnedquinn) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rex/Rex.kt:12]: What's the point of this? I'm thinking of `RexStruct` right now.

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/builder/RexBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        public fun lit(value: Long): RexBuilder = lit(Datum.bigint(value))
+
+        @JvmStatic
+        public fun lit(value: String): RexBuilder = lit(Datum.string(value))
+
+        @JvmStatic
+        public fun lit(value: Datum): RexBuilder = RexBuilder { it.rexLit(value) }
+
+        @JvmStatic
+        public fun collection(values: List<Rex>): RexBuilder = RexBuilder {
```
Comment (johnedquinn) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/builder/RexBuilder.kt:58]: Bag / Array ?

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/PartiQLPlan.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+public interface PartiQLPlan {
+
+    public fun getStatement(): Statement
+}
+
+/**
+ * TODO INTERNALIZE ME
+ * @property statement
+ */
+public class PartiQLPlanImpl(
```
Comment (johnedquinn) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/PartiQLPlan.kt:12]: Why does this need to be public in this PR?

**Review by johnedquinn — COMMENTED** [review]
Since I didn't see the initial PR for these, I've published a couple comments regarding the modeling.

If the intention is to solely focus on the builders, I really like these additions and I can approve. Though, I'd just want to make sure these comments don't go unaddressed in future iterations of the API.

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rel/RelFilter.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    private var _inputs: List<Rel>? = null
+    private var _predicate: Rex = predicate
+    private var _ordered: Boolean = input.isOrdered()
+
+    override fun getInput(): Rel = _input
+
+    override fun getInputs(): List<Rel> {
+        if (_inputs == null) {
+            _inputs = listOf(_input)
+        }
```
[the remark this answers is not in the document]
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rel/RelFilter.kt:40]: `getInputs()` and `getOperands()` make these data structures a more generic tree, but these could be removed and added later (if needed)

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rel/RelJoin.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
+package org.partiql.plan.v1.operator.rel
+
+import org.partiql.plan.v1.Schema
+import org.partiql.plan.v1.operator.rex.Rex
+
+/**
+ * TODO DOCUMENTATION
+ */
+public interface RelJoin : Rel {
```
[the remark this answers is not in the document]
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rel/RelJoin.kt:9]: ack, will add these

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rex/Rex.kt (hunk 1 above)
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rex/Rex.kt:12]: For the tree, but could be removed until needed..

> `getInputs()` and `getOperands()` make these data structures a more generic tree, but these could be removed and added later (if needed)

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/builder/RexBuilder.kt (hunk 2 above)
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/builder/RexBuilder.kt:58]: ack

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/PartiQLPlan.kt (hunk 3 above)
Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/PartiQLPlan.kt:12]: It technically doesn't need to be in _this_ PR

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rel/RelCorrelate.kt [hunk 6]
```diff
@@ -0,0 +1,70 @@
+package org.partiql.plan.v1.operator.rel
+
+import org.partiql.plan.v1.Schema
+
+/**
+ * Logical operator for nested-loop joins (correlated subqueries // lateral joins).
+ */
+public interface RelCorrelate : Rel {
```
Comment (johnedquinn) [inline partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rel/RelCorrelate.kt:8]:
```suggestion
public interface RelJoinLateral : Rel {
```

I might be missing something. Is this modeling drawn from somewhere else?

Does it make sense to combine correlated subqueries with lateral joins? When we execute a subquery (`ExprSelect`), the environment is passed to it, and we nest the input environment in the current environment. So, the modeling of `RexSelect` would probably act similarly, where we don't need a correlate rel to accomplish that.

Comment (RCHowell) [reply partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rel/RelCorrelate.kt:8]: https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/rel/core/Correlate.java

**Review by johnedquinn — APPROVED** [review]

**Pull request merged**